### PR TITLE
Make ci work on windows

### PR DESCRIPTION
**Pull request by 0dminnimda**

### CODE REVIEW

**Comment by 0dminnimda** [comment]
Finally 😃

**Comment by 0dminnimda** [comment]
So, okay, if I've already brought some attention to this, what should I do with the C5045 warnings?
@scoder, @da-woods, @mattip

File: Tools/ci-run.sh [hunk 1]
```diff
@@ -2,126 +2,208 @@
...
+    echo "Installing requirements [apt]"
+    sudo apt-add-repository -y "ppa:ubuntu-toolchain-r/test"
+    sudo apt update -y -q
+    sudo apt install -y -q ccache gdb python-dbg python3-dbg gcc-$GCC_VERSION || exit 1
+
+    if [[ $BACKEND_IS_CPP = true ]]; then
+      sudo apt install -y -q g++-$GCC_VERSION || exit 1
+    fi
+    sudo /usr/sbin/update-ccache-symlinks
+    echo "/usr/lib/ccache" >> $GITHUB_PATH # export ccache to path
```
Comment (mattip) [inline Tools/ci-run.sh:27]: why did you need to touch this?

Comment (0dminnimda) [reply Tools/ci-run.sh:27]: I just made it better and more readable, + we don't need to set up the compilers if we are testing the code style

**Comment by da-woods** [comment]
> what should I do with the C5045 warnings?

We should have a look at the code that is causing the warning - can you cut and paste an example snippet? Is it coming from user code or utility code?

Then we should decide if we think it's a useful warning (to us) or not.

If it's warning about something real then we should fix it (but probably in a separate PR).
If it's from utility code and is a useless warning then we can just disable it (e.g. https://github.com/cython/cython/pull/4317). This doesn't have to be done in this PR.

**Comment by 0dminnimda** [comment]
@da-woods 
so here's warning
> D:\a\cython\cython\TEST_TMP\2\compile\cpp\argdefault\argdefault.cpp(2573) : warning C5045: Compiler will insert Spectre mitigation for memory load if /Qspectre switch specified

**Comment by da-woods** [comment]
I'm more interested in what C++ code is causing the message. Are you able to run the test locally with `--no-cleanup` as an option, and then dig out the relevant lines from the `TEST_TMP` directory?

**Comment by 0dminnimda** [comment]
> Are you able to run the test locally with `--no-cleanup` as an option, and then dig out the relevant lines from the `TEST_TMP` directory?

@da-woods not sure if this is the same code because this code doesn't look very suspicious:

image


```cpp
        kind = __Pyx_PyUnicode_KIND(s1);
        if (kind != __Pyx_PyUnicode_KIND(s2)) {
            goto return_ne;
        }
        data1 = __Pyx_PyUnicode_DATA(s1);
        data2 = __Pyx_PyUnicode_DATA(s2);
        if (__Pyx_PyUnicode_READ(kind, data1, 0) != __Pyx_PyUnicode_READ(kind, data2, 0)) {
            goto return_ne;
        } else if (length == 1) {
            goto return_eq;
        } else {
            int result = memcmp(data1, data2, (size_t)(length * kind));
            #if PY_MAJOR_VERSION < 3
            Py_XDECREF(owned_ref);
            #endif
            return (equals == Py_EQ) ? (result == 0) : (result != 0);
        }
    } else if ((s1 == Py_None) & s2_is_unicode) {
        goto return_ne;
    } else if ((s2 == Py_None) & s1_is_unicode) {
        goto return_ne;
    } else {
        int result;
        PyObject* py_result = PyObject_RichCompare(s1, s2, equals);
        #if PY_MAJOR_VERSION < 3
        Py_XDECREF(owned_ref);
        #endif
        if (!py_result)
            return -1;
        result = __Pyx_PyObject_IsTrue(py_result);
        Py_DECREF(py_result);
        return result;
```

**Comment by 0dminnimda** [comment]
i'll just attach the file here, this is text since github does not allow attaching .cpp

[argdefault.txt](https://github.com/0dminnimda/cython/files/6900432/argdefault.txt)

**Comment by 0dminnimda** [comment]
more about the [C5045](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/c5045?view=msvc-160)

**Comment by da-woods** [comment]
I'm not sure I can track down the source of the C5045 errors. I'm not convinced `Wall` is needed for MSVC though.

**Comment by 0dminnimda** [comment]
Work on this PR continues in #2 for the reasons described there.

**Comment by 0dminnimda** [comment]
As it seems all of the important changes were merged in cython so, closing this one as well

**Comment by mattip** [comment]
:+1:
